### PR TITLE
Reftype update

### PR DESCRIPTION
**Pull request by yanaise**
Refactoring type references, no outside influence

---
---
_Release Notes_: 
---
_User Notifications_:

### CODE REVIEW

File: packages/adapter-utils/test/utils.test.ts [hunk 1]
```diff
@@ -2048,15 +2047,12 @@ describe('Test utils.ts', () => {
...
       const obj = new ObjectType({ elemID: new ElemID('a', 'elemID') })
       const objRef = createRefToElmWithValue(obj)
       expect(objRef.elemID).toEqual(obj.elemID)
-      expect(objRef.value).toEqual(obj)
-      const inst = new InstanceElement('ab', obj)
-      const instRef = createRefToElmWithValue(inst)
-      expect(instRef.elemID).toEqual(inst.elemID)
-      expect(instRef.value).toEqual(inst)
+      expect(objRef.type).toEqual(obj)
+      // TODO: Add tests for all possible TypeElements
```
Comment (adamro) [inline packages/adapter-utils/test/utils.test.ts:2055]: Not sure if super important but don't keep the TODO anyway :) [we know there's gonna be a rebase issue here with the other PR]

**Review by adamro — COMMENTED** [review]
LGTM but I want @roironn to review the serializer part here

File: packages/adapter-api/src/values.ts [hunk 2]
```diff
@@ -146,6 +146,23 @@ export class VariableExpression extends ReferenceExpression {
...
+export class TypeReference extends ReferenceExpression {
+  constructor(
+    public readonly elemID: ElemID,
+    public readonly type?: TypeElement,
+  ) {
+    super(elemID, type)
+  }
+
+  get traversalParts(): string[] {
+    return this.elemID.getFullNameParts()
```
Comment (roironn) [inline packages/adapter-api/src/values.ts:158]: why not use `super` here?

File: packages/adapter-api/src/values.ts [hunk 3]
```diff
@@ -146,6 +146,23 @@ export class VariableExpression extends ReferenceExpression {
   }
 }
 
+export class TypeReference extends ReferenceExpression {
+  constructor(
+    public readonly elemID: ElemID,
+    public readonly type?: TypeElement,
+  ) {
+    super(elemID, type)
```
Comment (roironn) [inline packages/adapter-api/src/values.ts:154]: adding a validation that the elemID here is not nested would be nice :)

File: packages/workspace/src/expressions.ts [hunk 4]
```diff
@@ -178,34 +178,30 @@ resolveTemplateExpression = async (
   }).toArray())
   .join('')
 
-const resolveElement = async (
-  elementToResolve: Element,
+const resolveElement = async <T extends Element>(
```
Comment (roironn) [inline packages/workspace/src/expressions.ts:181]: I can't add a comment at the proper line so adding here - when resolving a type reference we don't need to resolve the topLevelElement so this should be reflected in the code.

**Review by roironn — APPROVED** [review]
LGTM so approving (see comments before merging :) )

**Pull request merged**